File: frontend/src/types/expense.ts
import { Category, SubCategory, ExpenseType } from './category';

export interface Expense {
  id: string;
  category: Category;
  subCategory?: SubCategory;
  amount: number;
  description?: string;
  date: string;
  expenseType: ExpenseType;
  createdAt: string;
}

export interface CreateExpenseDTO {
  categoryId?: string;
  subCategoryId?: string;
  amount: number;
  description?: string;
  date: string;
  expenseType?: ExpenseType;
}

export interface ExpenseFilters {
  startDate?: string;
  endDate?: string;
  categoryId?: string;
  subCategoryId?: string;
  minAmount?: number;
  maxAmount?: number;
}

File: backend/src/main/java/com/houseexpenses/model/Expense.java
package com.houseexpenses.model;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "expenses")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Expense {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "user_id", nullable = false)
    private UUID userId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "subcategory_id")
    private SubCategory subCategory;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal amount;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "expense_date", nullable = false)
    private LocalDate expenseDate;

    @Enumerated(EnumType.STRING)
    @Column(name = "expense_type")
    @Builder.Default
    private ExpenseType expenseType = ExpenseType.monthly;

    @Column(name = "created_at")

    public enum ExpenseType {
        monthly, annual
    }
    private OffsetDateTime createdAt;

    @Column(name = "updated_at")
    private OffsetDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = OffsetDateTime.now();
        updatedAt = OffsetDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = OffsetDateTime.now();
    }
}
